Fix `typescript` dependency

We have a dependency on `typescript` via `@typescript-eslint/typescript-estree`. Since `typescript` isn't defined in the `@checkly/cli` `package.json`, this causes errors when users run the CLI. 

A simple solution would be to just add `typescript` in the `package.json`. Maybe we should make the dependency necessary only for projects that use `typescript`, though. Maybe it should be a peer dependency?


<details>
  <summary>Error message:</summary>
  
```
$ npx checkly test
(node:7952) [MODULE_NOT_FOUND] ModuleLoadError Plugin: @checkly/cli: [MODULE_NOT_FOUND] require failed to load /Users/clample/projects/nuxtjs/node_modules/@checkly/cli/dist/commands/deploy.js: Cannot find module 'typescript'
Require stack:
- /Users/clample/projects/nuxtjs/node_modules/@typescript-eslint/typescript-estree/dist/convert.js
- /Users/clample/projects/nuxtjs/node_modules/@typescript-eslint/typescript-estree/dist/ast-converter.js
- /Users/clample/projects/nuxtjs/node_modules/@typescript-eslint/typescript-estree/dist/parser.js
- /Users/clample/projects/nuxtjs/node_modules/@typescript-eslint/typescript-estree/dist/index.js
- /Users/clample/projects/nuxtjs/node_modules/@checkly/cli/dist/services/check-dependency-parser.js
- /Users/clample/projects/nuxtjs/node_modules/@checkly/cli/dist/constructs/browser-check.js
- /Users/clample/projects/nuxtjs/node_modules/@checkly/cli/dist/constructs/index.js
- /Users/clample/projects/nuxtjs/node_modules/@checkly/cli/dist/services/project-parser.js
- /Users/clample/projects/nuxtjs/node_modules/@checkly/cli/dist/commands/deploy.js
- /Users/clample/projects/nuxtjs/node_modules/@oclif/core/lib/module-loader.js
- /Users/clample/projects/nuxtjs/node_modules/@oclif/core/lib/config/plugin.js
- /Users/clample/projects/nuxtjs/node_modules/@oclif/core/lib/config/config.js
- /Users/clample/projects/nuxtjs/node_modules/@oclif/core/lib/config/index.js
- /Users/clample/projects/nuxtjs/node_modules/@oclif/core/lib/command.js
- /Users/clample/projects/nuxtjs/node_modules/@oclif/core/lib/index.js
- /Users/clample/projects/nuxtjs/node_modules/@checkly/cli/bin/run
module: @oclif/core@1.21.0
task: toCached
plugin: @checkly/cli
root: /Users/clample/projects/nuxtjs/node_modules/@checkly/cli
See more details with DEBUG=*
(Use `node --trace-warnings ...` to show where the warning was created)
(node:7952) [MODULE_NOT_FOUND] ModuleLoadError Plugin: @checkly/cli: [MODULE_NOT_FOUND] require failed to load /Users/clample/projects/nuxtjs/node_modules/@checkly/cli/dist/commands/test.js: Cannot find module 'typescript'
Require stack:
- /Users/clample/projects/nuxtjs/node_modules/@typescript-eslint/typescript-estree/dist/convert.js
- /Users/clample/projects/nuxtjs/node_modules/@typescript-eslint/typescript-estree/dist/ast-converter.js
- /Users/clample/projects/nuxtjs/node_modules/@typescript-eslint/typescript-estree/dist/parser.js
- /Users/clample/projects/nuxtjs/node_modules/@typescript-eslint/typescript-estree/dist/index.js
- /Users/clample/projects/nuxtjs/node_modules/@checkly/cli/dist/services/check-dependency-parser.js
- /Users/clample/projects/nuxtjs/node_modules/@checkly/cli/dist/constructs/browser-check.js
- /Users/clample/projects/nuxtjs/node_modules/@checkly/cli/dist/constructs/index.js
- /Users/clample/projects/nuxtjs/node_modules/@checkly/cli/dist/services/project-parser.js
- /Users/clample/projects/nuxtjs/node_modules/@checkly/cli/dist/commands/test.js
- /Users/clample/projects/nuxtjs/node_modules/@oclif/core/lib/module-loader.js
- /Users/clample/projects/nuxtjs/node_modules/@oclif/core/lib/config/plugin.js
- /Users/clample/projects/nuxtjs/node_modules/@oclif/core/lib/config/config.js
- /Users/clample/projects/nuxtjs/node_modules/@oclif/core/lib/config/index.js
- /Users/clample/projects/nuxtjs/node_modules/@oclif/core/lib/command.js
- /Users/clample/projects/nuxtjs/node_modules/@oclif/core/lib/index.js
- /Users/clample/projects/nuxtjs/node_modules/@checkly/cli/bin/run
module: @oclif/core@1.21.0
task: toCached
plugin: @checkly/cli
root: /Users/clample/projects/nuxtjs/node_modules/@checkly/cli
See more details with DEBUG=*
 ›   Error: command test not found
 ```
 </details>

--- a/README.md
+++ b/README.md
@@ -21,6 +21,12 @@ First, install the CLI.
 npm i --save-dev @checkly/cli 
 ```
 
+To use TypeScript, also install `ts-node` and `typescript`:
+
+```bash
+npm i --save-dev ts-node typescript
+```
+
 Create a `checkly.config.js` (or `checkly.config.ts`) at the root of your project.
 
 ```js

--- a/package/src/services/check-dependency-parser.ts
+++ b/package/src/services/check-dependency-parser.ts
@@ -1,9 +1,8 @@
 import * as path from 'path'
 import * as fs from 'fs'
 import * as acorn from 'acorn'
-import * as tsParser from '@typescript-eslint/typescript-estree'
-import { TSESTree } from '@typescript-eslint/typescript-estree'
 import * as walk from 'acorn-walk'
+import { TSESTree } from '@typescript-eslint/typescript-estree'
 
 type UnsupportedNpmDependencies = {
   file: string;
@@ -154,11 +153,12 @@ function parseDependenciesForFile (filePath: string): { localDependencies: strin
     const ast = acorn.parse(contents, { allowReturnOutsideFunction: true, ecmaVersion: 'latest' })
     walk.simple(ast, jsNodeVisitor(localDependencies, npmDependencies))
   } else if (filePath.endsWith('.ts')) {
+    const tsParser = getTsParser()
     const ast = tsParser.parse(contents, {})
     // The AST from typescript-estree is slightly different from the type used by acorn-walk.
     // This doesn't actually cause problems (both are "ESTree's"), but we need to ignore type errors here.
     // @ts-ignore
-    walk.simple(ast, tsNodeVisitor(localDependencies, npmDependencies))
+    walk.simple(ast, tsNodeVisitor(tsParser, localDependencies, npmDependencies))
   } else {
     throw new Error(`Unsupported file extension for ${filePath}`)
   }
@@ -176,18 +176,18 @@ function jsNodeVisitor (localDependencies: Set<string>, npmDependencies: Set<str
   }
 }
 
-function tsNodeVisitor (localDependencies: Set<string>, npmDependencies: Set<string>): any {
+function tsNodeVisitor (tsParser: any, localDependencies: Set<string>, npmDependencies: Set<string>): any {
   return {
     ImportDeclaration (node: TSESTree.ImportDeclaration) {
       // For now, we only support literal strings in the import statement
-      if (node.source.type !== TSESTree.AST_NODE_TYPES.Literal) return
+      if (node.source.type !== tsParser.TSESTree.AST_NODE_TYPES.Literal) return
       registerDependency(node.source.value, localDependencies, npmDependencies)
     },
     ExportNamedDeclaration (node: TSESTree.ExportNamedDeclaration) {
       // The statement isn't importing another dependency
       if (node.source === null) return
       // For now, we only support literal strings in the import statement
-      if (node.source.type !== TSESTree.AST_NODE_TYPES.Literal) return
+      if (node.source.type !== tsParser.TSESTree.AST_NODE_TYPES.Literal) return
       registerDependency(node.source.value, localDependencies, npmDependencies)
     },
   }
@@ -239,4 +239,19 @@ function registerDependency (importArg: string | null, localDependencies: Set<st
   } else {
     npmDependencies.add(importArg)
   }
+}
+
+// To avoid a dependency on typescript for users with no TS checks, we need to dynamically import typescript-estree
+let tsParser: any
+function getTsParser (): any {
+  if (tsParser) return tsParser
+  try {
+    tsParser = require('@typescript-eslint/typescript-estree')
+  } catch (err: any) {
+    if (err.code === 'ERR_MODULE_NOT_FOUND' || err.code === 'MODULE_NOT_FOUND') {
+      throw new Error('Please install typescript to use TypeScript in check files')
+    }
+    throw err
+  }
+  return tsParser
 }

--- a/package/src/services/util.ts
+++ b/package/src/services/util.ts
@@ -41,6 +41,7 @@ export async function loadTsFile (filepath: string): Promise<any> {
   return exported
 }
 
+// To avoid a dependency on typescript for users with no TS checks, we need to dynamically import ts-node
 let tsCompiler: Service
 async function getTsCompiler (): Promise<Service> {
   if (tsCompiler) return tsCompiler
@@ -52,9 +53,10 @@ async function getTsCompiler (): Promise<Service> {
       },
     })
   } catch (err: any) {
-    if (err.code === 'ERR_MODULE_NOT_FOUND') {
+    if (err.code === 'ERR_MODULE_NOT_FOUND' || err.code === 'MODULE_NOT_FOUND') {
       throw new Error('Please install ts-node and typescript to use TypeScript configuration files')
     }
+    throw err
   }
   return tsCompiler
 }

